File: api/cplus/src/router.cpp
/**
 * Tencent is pleased to support the open source community by making Tseer available.
 *
 * Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 * 
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed 
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

#include "router.h"

#include "global.h"
#include <iostream>
#include <sstream>
using std::ostringstream;
using std::string;
using std::vector;
using Tseer::RouterNodeInfo;

namespace Tseerapi
{

Router::Router(LB_TYPE type)
: _expreTime(0)
, _lastVisit(0)
, _lb(NULL)
{
    _lb = LoadBalanceFactory::CreateLoadBalance(type);
}

Router::~Router()
{
    if(_lb)
    {
        delete _lb;
        _lb = NULL;
    }
}
int Router::getRouters(vector<RouterNodeInfo>& nodeInfoVec, string &errMsg)
{
    size_t now = time(NULL);

    if(_expreTime < now)
    {
        int ret = _lb->getRouters(nodeInfoVec, errMsg);
        if(ret == 0)
        {
            return -2;
        }

        return -1;
    }
    _lastVisit = now;

    return _lb->getRouters(nodeInfoVec, errMsg);
}

int Router::getRouter(RouterNodeInfo& nodeInfo, string &errMsg)
{
    size_t now = time(NULL);

    if(_expreTime < now)
    {
        int ret = _lb->getRouter(nodeInfo, errMsg);
        if(ret == 0)
        {
            return -2;
        }

        return -1;
    }
    _lastVisit = now;

    return _lb->getRouter(nodeInfo, errMsg);
}

int Router::getRouter(unsigned long long key, RouterNodeInfo& nodeInfo, string &errMsg)
{
    size_t now = time(NULL);

    if(_expreTime < now)
    {
        int ret = _lb->getRouter(key, nodeInfo, errMsg);
        if(ret == 0)
        {
            return -2;
        }

        return -1;
    }
    _lastVisit = now;

    return _lb->getRouter(key, nodeInfo, errMsg);
}

int Router::updateRouterNodes(const vector<RouterNodeInfo> &RouterNodeInfoVec, string &errMsg)
{
    if(_lb == NULL)
    {
        ostringstream os;
        os << FILE_FUN << "lb is not created";
        errMsg = os.str();
        return -1;
    }

    _lb->clear();
    _lb->del(RouterNodeInfoVec);

    for(size_t i = 0; i < RouterNodeInfoVec.size(); ++i)
    {
        int ret = _lb->addRouter(RouterNodeInfoVec[i]);
        if(ret != 0)
        {
            ostringstream os;
            os << FILE_FUN << "add node error";
            errMsg = os.str();
            return ret;
        }
    }

    if(_lb->rebuild())
    {
        ostringstream os;
        os << FILE_FUN << "rebuild node error";
        errMsg = os.str();
        return -1;
    }

return 0;
}

int Router::statReqResult(const InnerRouterRequest& req, int ret, int timeCost)
{
    RouterNodeInfo nodeInfo;
    nodeInfo.ip = req.ip;
    nodeInfo.port = req.port;
    nodeInfo.isTcp = req.isTcp;

    if (req.isAgent)
    {
        return _statMgr.reportStat(req, ret, timeCost);
    }
    else
    {
        return _lb->statResult(nodeInfo, ret, timeCost);
    }
}

void Router::updateExpretime(time_t interval)
{
    size_t now = time(NULL);
    _expreTime = now + interval;
}

}

File: api/java/src/main/java/com/qq/cloud/router/client/Router.java
/**
 * Tencent is pleased to support the open source community by making Tseer available.
 *
 * Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 * 
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * 
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed 
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.qq.cloud.router.client;

public interface Router {

	/**
	 * 初始化路由，如果无特殊需求不需要手动调用
	 * @param cfg 初始化配置
	 */
    void init(RouterConfig cfg);

    /**
     * 获取单个节点
     * @param req 路由请求
     * @return 包含目标节点的ip,port等信息和返回描述信息
     */
    Result<RouterResponse> getRouter(RouterRequest req);
    
    /**
     * 获取所有目标节点
     * @param req 路由请求
     * @return 目标节点列表
     */
    Result<RoutersResponse> getRouters(RoutersRequest req);

    /**
     * 模调数据上报接口
     * @param req 路由请求
     * @param res 路由回复
     * @param ret 调用被调服务的返回值
     * @param timecost 调用被调服务花费的时间，单位毫秒
     * @return 描述信息
     */
    Result<String> resultReport(RouterRequest req, RouterResponse res, int ret, long timecost);
}
